File: leetcode/binary-tree-preorder-traversal/binary-tree-preorder-traversal-iter-alt.rb
# LeetCode #144 - Binary Tree Preorder Traversal
# https://leetcode.com/problems/binary-tree-preorder-traversal/
# Alternative iterative solution.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def preorder_traversal(root)
  out = []
  stack = []

  while root || !stack.empty?
    # Go left as far as possible, visiting each node.
    while root
      out << root.val
      stack << root
      root = root.left
    end

    # If it is possible to go right, we'll do that next.
    root = stack.pop.right
  end

  out
end

File: leetcode/binary-tree-preorder-traversal/binary-tree-preorder-traversal-iter-alt.java
// LeetCode #114 - Binary Tree Preorder Traversal
// https://leetcode.com/problems/binary-tree-preorder-traversal/
// Alternative iterative implementation.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        preorder(root, values::add);
        return values;
    }

    private static void preorder(TreeNode root, IntConsumer action) {
        var stack = Collections.asLifoQueue(new ArrayDeque<TreeNode>());

        while (root != null || !stack.isEmpty()) {
            // Go left as far as possible, doing the preorder action.
            for (; root != null; root = root.left) {
                action.accept(root.val);
                stack.add(root);
            }

            // If we can go right, do that next.
            root = stack.remove().right;
        }
    }
}
